File: src/com/intellij/idea/plugin/hybris/type/system/meta/impl/TSMetaEnumImpl.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.type.system.meta.impl;

import com.intellij.idea.plugin.hybris.type.system.meta.TSMetaEnum;
import com.intellij.idea.plugin.hybris.type.system.meta.TSMetaEnumValue;
import com.intellij.idea.plugin.hybris.type.system.meta.impl.CaseInsensitive.NoCaseMultiMap;
import com.intellij.idea.plugin.hybris.type.system.model.EnumType;
import com.intellij.idea.plugin.hybris.type.system.model.EnumValue;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.stream.Stream;

public class TSMetaEnumImpl extends TSMetaEntityImpl<EnumType> implements TSMetaEnum {

    private final NoCaseMultiMap<TSMetaEnumValueImpl> name2ValueObj = new NoCaseMultiMap<>();
    private final boolean myAutocreate;
    private final boolean myGenerate;
    private final boolean myDynamic;
    private final String myDescription;
    private final String myJaloclass;

    public TSMetaEnumImpl(final Project project, final String name, final EnumType dom) {
        super(project, name, dom);
        myAutocreate = Boolean.TRUE.equals(dom.getAutoCreate().getValue());
        myGenerate = Boolean.TRUE.equals(dom.getGenerate().getValue());
        myDynamic = Boolean.TRUE.equals(dom.getDynamic().getValue());
        myDescription = dom.getDescription().getStringValue();
        myJaloclass = dom.getJaloClass().getStringValue();
    }

    @NotNull
    @Override
    public Stream<? extends TSMetaEnumValue> getValuesStream() {
        return name2ValueObj.values().stream();
    }

    @NotNull
    @Override
    public Collection<? extends TSMetaEnumValue> findValueByName(@NotNull final String name) {
        return new ArrayList<>(name2ValueObj.get(name));
    }

    @Override
    public void createValue(final @NotNull EnumValue domEnumValue) {
        final TSMetaEnumValueImpl result = new TSMetaEnumValueImpl(getProject(), this, domEnumValue);

        if (result.getName() != null) {
            name2ValueObj.putValue(result.getName(), result);
        }
    }

    @Override
    public boolean isAutocreate() {
        return myAutocreate;
    }

    @Override
    public boolean isGenerate() {
        return myGenerate;
    }

    @Override
    public boolean isDynamic() {
        return myDynamic;
    }

    @Override
    public String getDescription() {
        return myDescription;
    }

    @Override
    public String getJaloclass() {
        return myJaloclass;
    }
}

File: src/com/intellij/idea/plugin/hybris/type/system/meta/model/impl/TSMetaEnumImpl.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.type.system.meta.model.impl

import com.intellij.idea.plugin.hybris.type.system.meta.impl.CaseInsensitive.NoCaseMultiMap
import com.intellij.idea.plugin.hybris.type.system.meta.model.TSMetaEnum
import com.intellij.idea.plugin.hybris.type.system.meta.model.TSMetaEnum.TSMetaEnumValue
import com.intellij.idea.plugin.hybris.type.system.model.EnumType
import com.intellij.idea.plugin.hybris.type.system.model.EnumValue
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.util.xml.DomAnchor
import com.intellij.util.xml.DomService

class TSMetaEnumImpl(
    override val module: Module,
    override val project: Project,
    override val name: String?,
    dom: EnumType,
    override val isCustom: Boolean
) : TSMetaEntityImpl<EnumType>(dom, module, project, isCustom, name), TSMetaEnum {

    private val myAllDoms: MutableSet<DomAnchor<EnumType>> = LinkedHashSet()
    override val values = NoCaseMultiMap<TSMetaEnumValue>()
    override val isAutoCreate = java.lang.Boolean.TRUE == dom.autoCreate.value
    override val isGenerate = java.lang.Boolean.TRUE == dom.generate.value
    override val isDynamic = java.lang.Boolean.TRUE == dom.dynamic.value
    override val description = dom.description.stringValue
    override val jaloClass = dom.jaloClass.stringValue

    init {
        myAllDoms.add(DomService.getInstance().createAnchor(dom))
    }

    override fun findValueByName(name: String) = ArrayList(values[name])
    override fun retrieveAllDomsStream() = myAllDoms.mapNotNull { it.retrieveDomElement() }

    override fun merge(another: TSMetaEnum) {
        another.values.values().stream()
            .filter { anotherValue -> anotherValue.name != null }
            .forEach { anotherValue -> values.putValue(anotherValue.name!!, anotherValue) }
        addDomRepresentation(another.retrieveDom()!!)
    }

    private fun addDomRepresentation(anotherDom: EnumType) =
        myAllDoms.add(DomService.getInstance().createAnchor(anotherDom))

    class TSMetaEnumValueImpl(
        override val module: Module,
        override val project: Project,
        dom: EnumValue,
        override val isCustom: Boolean,
        override val owner: TSMetaEnum,
        override val name: String
    ) : TSMetaEntityImpl<EnumValue>(dom, module, project, isCustom, name),
        TSMetaEnumValue {

        override val description = dom.description.stringValue

    }
}